packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,477 @@
+   * the asset being bought.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * Assumes both CELO and the stable token have 18 decimals.
+   * @param stableToken The stableToken involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The amount of the asset being bought.
+   */
+  function getBuyAmount(address stableToken, uint256 sellAmount, bool sellCelo)
+    public
+    view
+    returns (uint256)
+  {
+    // Gets the price of CELO quoted in stableToken.
+    FixidityLib.Fraction memory exchangeRate = getOracleExchangeRate(stableToken);
+    // If stableToken is being sold, instead use the price of stableToken
+    // quoted in CELO.
+    if (!sellCelo) {
+      exchangeRate = exchangeRate.reciprocal();
+    }
+    // The sell amount taking the spread into account, ie:
+    // (1 - spread) * sellAmount
+    FixidityLib.Fraction memory adjustedSellAmount = FixidityLib.fixed1().subtract(spread).multiply(
+      FixidityLib.newFixed(sellAmount)
+    );
+    // Calculate the buy amount:
+    // exchangeRate * adjustedSellAmount
+    return exchangeRate.multiply(adjustedSellAmount).fromFixed();
+  }
+
+  /**
+   * @notice Gets the proposal identifiers of all exchange proposals in the
+   * Proposed or Approved state.
+   * @return An array of active exchange proposals IDs.
+   */
+  function getActiveProposalIds() external view returns (uint256[] memory) {
+    return activeProposalIds.getKeys();
+  }
+
+  /**
+   * @notice Gets the oracle CELO price quoted in the stable token.
+   * @dev Reverts if there is not a rate for the provided stable token.
+   * @param stableToken The stable token to get the oracle price for.
+   * @return The oracle CELO price quoted in the stable token.
+   */
+  function getOracleExchangeRate(address stableToken)
+    private
+    view
+    returns (FixidityLib.Fraction memory)
+  {
+    uint256 rateNumerator;
+    uint256 rateDenominator;
+    (rateNumerator, rateDenominator) = getSortedOracles().medianRate(stableToken);
+    // When rateDenominator is 0, it means there are no rates known to SortedOracles.
+    require(rateDenominator > 0, "No oracle rates present for token");
+    return FixidityLib.wrap(rateNumerator).divide(FixidityLib.wrap(rateDenominator));
+  }
+
+  /**
+   * @notice Sets the approver.
+   * @dev Sender must be owner. New approver is allowed to be address(0).
+   * @param newApprover The new value for the spread.
+   */
+  function setApprover(address newApprover) public onlyOwner {
+    approver = newApprover;
+    emit ApproverSet(newApprover);
+  }
+
+  /**
+   * @notice Sets the spread.
+   * @dev Sender must be owner.
+   * @param newSpread The new value for the spread.
+   */
+  function setSpread(uint256 newSpread) public onlyOwner {
+    spread = FixidityLib.wrap(newSpread);
+    emit SpreadSet(newSpread);
+  }
```

Comment:
in the past we have required `<= fixed1` on fractions like this
it should be explicitly documented that the input is wrapped within the function (in @ dev ?)
or just shouldn't be wrapped; imo whatever makes the config values easier to read in plaintext should be prioritized

Reply:
ignore the latter half, forgot what `wrap` meant lol